### ISSUE TITLE
Advanced Maven resolution strategies

### ISSUE THREAD

**Issue by spyrkob** [wildfly-extras/wildfly-channel#97]
Currently channels require each distinct groupId to be listed separately. In case of open channels, limiting versions to known pattern, it means a long list of streams with the same artifactId (`*`) and versionPattern (eg. `.*.Final`).

It would be easier to be define a catch-all stream that would be applied to all streams not explicitly defined. For example:
```
- groupId: '*'
   artifactId: '*'
   versionPattern: '.*.Final'
```

**Comment by jmesnil** [comment]
I can see the benefit of a catch-them-all stream to provide the latest version of any component but in practice, this should be avoided as it does not provide any curation benefit unless you tightly control the Maven repositories that are queried.

The `versionPattern: '.*.Final'` would not be correct and would miss many components that are included in WildFly. We likely could do no better than `versionPattern: '.*'`.

At this point, I wonder if the should not have a special field in the channel to define the resolution of unspecified streams.

If the channel does not define a stream matching the GA of a component:
* do nothing and use the specified version when ChannelSession is called (by default)
* or use the latest release *according* to Maven metadata

In the channel.yaml file, we could have a top field such as 

```
resolves-if-no-stream: latest | release | original # default is original
``` 

* if `resolves-if-no-stream: original`, we would fetch the Maven GAV passed to the ChannelSession API
* if `resolves-if-no-stream: latest `, we would fetch the Maven GAV corresponding to the `<versioning><latest>` from the Maven metadata.
* if `resolves-if-no-stream: release `, we would fetch the Maven GAV corresponding to the `<versioning><release>` from the Maven metadata.

**Comment by spyrkob** [comment]
Yes the version Pattern would have to be very generic - I was mostly thinking of product side where we can filter by the suffix. On the other hand in that scenario we're using a repository we don't expect to contain artifacts without that suffix, so it's not much of filtering 😉 

I think `resolves-if-no-stream` would fit the use case I had in mind... I'll try and implement that next week

Should it also have an option like `none` to prevent resolution artifacts not explicitly listed in the stream?

**Comment by spyrkob** [comment]
Fixed by #109
